### ISSUE TITLE
Clarification on the mutual exclusivity/precendence of `:other`, `:align`, `:ratio`

### ISSUE THREAD

**Issue by kaushalmodi** [wasamasa/shackle#20]
I tried the below and it did not work:

**Snippet 1**
```el
(setq shackle-rules
      '(("*Help*" :select t :ratio 0.3 :align right :other t))
```

The \*Help\* buffer was always opened in a new window.

But the below, as suggested by you, worked:

**Snippet 2**
```el
(setq shackle-rules
      '(("*Help*" :select t :other t))
```

So the questions I have are:

1. If I use `:ratio` and `:align`, will a new window be popped, always?
2. How can I get the `:other` functionality in addition to controlling the `:ratio` of a new popped up window (if a new window needs to be created, as in the `(one-window-p)` case)?
3. Should `:ratio` and `:align` be not used with `:other`?

**Comment by wasamasa** [comment]
1: Yes, the point of `:align` is to always display a window occupying the full width/height of the frame, `:ratio` merely dictates the percentage of the frame it shall use.

2: I'm not sure I get your question. A window as created by `:align` cannot be a reused window by design, so `:other` wouldn't make any sense for it. Perhaps some crude ASCII art will explain their nature:

```
+-------+    +-------+
|   A   |    |   A   |
|       |    +-------+
+-------+ -> |   B   |
|   B   |    +-------+
|       |    |   C   |
+-------+    +-------+

+---+---+    +---+---+
|   |   |    |   |   |
|   |   |    | A | B |
| A | B | -> |   |   |
|   |   |    +---+---+
|   |   |    |   C   |
+---+---+    +-------+
```

Here's what happens if you use `:align` on a horizontal and vertical split.  As you can see, the window C always gets displayed at the bottom.  This is done by unconditionally splitting the root window which encompasses both window A and B.

3: That's right, `:align` and `:other` are mutually exclusive. So far I've relied on common sense to explain the interactions (all keys are exclusive unless indicated otherwise in the docs), a table illustrating how they interact would be probably overkill or not even properly doable (as there are lots of invalid combinations and relatively few valid ones).

**Comment by kaushalmodi** [comment]
Thanks for the detailed explaination. My reasoning for **Snippet 1** was to achieve the below:

```el
(setq shackle-rules
      '(("*Help*" :select t :ratio 0.3 :align right :other t))
```

- If one window, new window splits on the right with 0.3 ratio, move point to new window
```
+----------+    +------+---+
|          |    |      |*  |
|    *     |    |      |h  |
|    A     | -> |  A   |e  |
|          |    |      |l  |
|          |    |      |p  |
+----------+    +------+---+
```

- If >1 windows, reuse an existing window (don't change ratio), move point to new window
```
+----+-----+    +----+-----+
|    |     |    |    |  *  |
|  * |     |    |    |  h  |
|  A |  B  | -> |  A |  e  |
|    |     |    |    |  l  |
|    |     |    |    |  p  |
+----+-----+    +----+-----+```

**Comment by wasamasa** [comment]
I'm a bit reluctant on implementing that specific interaction as it isn't immediately obvious that the combination of `:align` and `:other` will result in your described behaviour. I'd be happier making `:ratio` work for other types than `:align` if you don't need the behaviour of it splitting the root window to span the entire frame.

**Comment by kaushalmodi** [comment]
This might need quite from rework but here's the idea:
- `:ratio` and `:align` should only be effective if a "window popup" has to happen
- Window popup can happen due to any of these reasons:
  - Use of `:popup t`
  - Use of `:same nil`
  - User used `:other` and there was only one window when `display-buffer` was called and thus the required popup action

>  I'd be happier making `:ratio` work for other types than `:align` if you don't need the behaviour of it splitting the root window to span the entire frame.

I didn't understand that. `:other` is already doing a good job of splitting the root window if only one window is present. The only thing I miss is that `:ratio` and `:align` are not effective on the window popup caused by `:other`.

In other words, would it be better if `:same`, `:popup`, `:other` dictated the top-level window behavior; and then `:align` and `:ratio` come into play only when an actual window popup is needed? If you decide to do that, then they as well be called `:pop-align` and `:pop-ratio` instead.

**Comment by wasamasa** [comment]
`:other` does *not* split the root window, it just changes the default behaviour of popping up a window (which can be enforced by using `:popup`).

What you're proposing is mixing mutually exclusive behaviours depending on the number of windows which does complicate things needlessly for little benefit. I see the need for making the documentation more clear, for instance by declaring `:same`, `:popup` and `:other` as modifiers and `:align` or `:frame` as an alternative to the normal `display-buffer` behaviour that would be otherwise active.

**Comment by kaushalmodi** [comment]
Sorry, I meant that the use of `:other t` option is responsible for the popup to happen in the case of one-window frame.

I am alright with current state of the package and not being able to customize the ratio of the popped up window when using `:other`. More than that I am happy that I got a hang of how these keywords are related. Thanks for the explanations.
